<img src="../../predictioNN_Logo_JPG(72).jpg" width=200>

---

## Logistic Regression with Breast Cancer Data

### Introduction to Data Science
#### Last Updated: February 27, 2023
---  

### SOURCES
- [Logistic regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)

### OBJECTIVES
- Implement logistic regression using `sklearn`
- Use the sigmoid function to compute the predicted probability
- Compute binary classification metrics

### CONCEPTS
- logistic regression

---


## 1. Logistic Regression with `sklearn`

We worked with the sigmoid function for computing probabilities of a binary outcome.  
Logistic regression is a model that does these things:
- Use a linear combination of predictors as input, equal to: $\beta_0 + \beta_1 X_1 + ... + \beta_n X_n$
- Feed the input into the sigmoid function (a.k.a. logistic function)
- Estimate the parameters to minimize error
- Output a probability estimate of the outcome

Now we will work with the Wisconsin Breast Cancer Dataset to predict if a cell is benign ('B') or malignant ('M'). 

The dataset was sourced here:  
https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(diagnostic)

In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [2]:
datapath = '../datasets/wdbc.csv'

**Read in the Data**

In [3]:
df = pd.read_csv(datapath)
df.head()

id diagnosis     f1     f2      f3      f4       f5       f6      f7  \
0    842302         M  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.3001   
1    842517         M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.0869   
2  84300903         M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.1974   
3  84348301         M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.2414   
4  84358402         M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.1980   

        f8  ...    f21    f22     f23     f24     f25     f26     f27     f28  \
0  0.14710  ...  25.38  17.33  184.60  2019.0  0.1622  0.6656  0.7119  0.2654   
1  0.07017  ...  24.99  23.41  158.80  1956.0  0.1238  0.1866  0.2416  0.1860   
2  0.12790  ...  23.57  25.53  152.50  1709.0  0.1444  0.4245  0.4504  0.2430   
3  0.10520  ...  14.91  26.50   98.87   567.7  0.2098  0.8663  0.6869  0.2575   
4  0.10430  ...  22.54  16.67  152.20  1575.0  0.1374  0.2050  0.4000  0.1625   

      f29      f30  
0  0.4601  0.11890  
1  0.2750  0.08902  
2  0.3613  0.08758  
3  0.6638  0.17300  
4  0.2364  0.07678  

[5 rows x 32 columns]

**About the Fields**

`id` - unique identifier for each subject  
`diagnosis` - target variable indicating if cell is malignant or benign  
`f1-f30` - cell measurement variables

---

**Preprocessing**

The `diagnosis` field is the target variable. It needs to be converted to values of 0 and 1.  
We can make malignant = 1, benign = 0.

In [5]:
df['target'] = df['diagnosis'].apply(lambda x: 1 if x == 'M' else 0)

Let's make sure this is correct

In [6]:
print(df.head())
print(df.tail())

         id diagnosis     f1     f2      f3      f4       f5       f6      f7  \
0    842302         M  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.3001   
1    842517         M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.0869   
2  84300903         M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.1974   
3  84348301         M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.2414   
4  84358402         M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.1980   

        f8  ...    f22     f23     f24     f25     f26     f27     f28  \
0  0.14710  ...  17.33  184.60  2019.0  0.1622  0.6656  0.7119  0.2654   
1  0.07017  ...  23.41  158.80  1956.0  0.1238  0.1866  0.2416  0.1860   
2  0.12790  ...  25.53  152.50  1709.0  0.1444  0.4245  0.4504  0.2430   
3  0.10520  ...  26.50   98.87   567.7  0.2098  0.8663  0.6869  0.2575   
4  0.10430  ...  16.67  152.20  1575.0  0.1374  0.2050  0.4000  0.1625   

      f29      f30  target  
0  0.4601  0.11890       1  
1  0.2750 

**Prepare the data, using f1 and f2 as predictors**

In [7]:
X = df[['f1','f2']].values
y = df['target'].values

**Split into training set, test set**

In [8]:
x_tr, x_te, y_tr, y_te = train_test_split(X, y, train_size = 0.6, random_state=314)

**Print some of the data**

In [9]:
print('x_tr: \n', x_tr[:5,:])
print('')
print('y_tr: \n', y_tr[:5])

x_tr: 
 [[19.21 18.57]
 [19.59 25.  ]
 [10.29 27.61]
 [13.85 19.6 ]
 [12.47 18.6 ]]

y_tr: 
 [1 1 0 0 0]


**Fit the logistic regression model**

Turn off penality for large parameters

In [178]:
model = LogisticRegression(penalty='none').fit(x_tr, y_tr)

**Extract the model coefficients**

In [179]:
model.coef_

array([[1.10417367, 0.22893814]])

**Extract the model intercept**

In [180]:
model.intercept_

array([-20.75395288])

This tells us that the coefficient on f1 and f2 is 1.10417367 and 0.22893814, respectively

**Predict the Cell Types**

In [181]:
# store the predictions for all subjects in a new column
df['label_predicted'] = model.predict(X)

# print the predictions
print(model.predict(X))

[1 1 1 0 1 0 1 0 0 0 1 1 1 1 0 1 1 1 1 0 0 0 0 1 1 1 1 1 1 1 1 0 1 1 1 1 0
 0 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 0 1 0 0 1 0 0 0 0 1 0 1 0
 0 1 0 1 1 0 0 0 1 1 0 1 1 1 0 0 1 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 0 0 1 0 0 0 0 0 1
 0 0 0 0 0 0 0 0 1 1 0 0 0 1 1 0 1 1 0 1 1 0 0 0 0 0 0 0 0 1 0 0 1 1 1 0 1
 0 1 0 0 0 1 0 0 0 1 0 0 1 1 0 0 1 1 0 0 0 0 1 0 0 1 0 1 1 1 0 0 0 1 1 0 0
 0 1 0 0 0 0 0 0 1 0 0 1 0 0 1 1 1 1 0 0 0 0 1 0 0 0 0 0 1 0 1 1 1 0 1 1 1
 1 1 1 1 1 1 1 0 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0
 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 1 1 1 0 0
 0 0 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 1 0 1 1 0 1 1
 1 0 1 1 0 1 0 1 0 0 0 0 0 0 0 1 0 0 0 1 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 1
 0 1 0 0 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0
 1 0 1 0 0 1 0 1 0 0 0 1 0 0 0 0 1 1 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 1 0 1 0 1 1 

**Filter the dataframe to show subjects where the prediction matches the target. These are correct predictions.**

In [182]:
df[df['label_predicted'] == df['target']]

id diagnosis     f1     f2      f3      f4       f5       f6  \
0      842302         M  17.99  10.38  122.80  1001.0  0.11840  0.27760   
1      842517         M  20.57  17.77  132.90  1326.0  0.08474  0.07864   
2    84300903         M  19.69  21.25  130.00  1203.0  0.10960  0.15990   
4    84358402         M  20.29  14.34  135.10  1297.0  0.10030  0.13280   
6      844359         M  18.25  19.98  119.60  1040.0  0.09463  0.10900   
..        ...       ...    ...    ...     ...     ...      ...      ...   
564    926424         M  21.56  22.39  142.00  1479.0  0.11100  0.11590   
565    926682         M  20.13  28.25  131.20  1261.0  0.09780  0.10340   
566    926954         M  16.60  28.08  108.30   858.1  0.08455  0.10230   
567    927241         M  20.60  29.33  140.10  1265.0  0.11780  0.27700   
568     92751         B   7.76  24.54   47.92   181.0  0.05263  0.04362   

          f7       f8  ...     f23     f24      f25      f26     f27     f28  \
0    0.30010  0.14710  ...  184.60  2019.0  0.16220  0.66560  0.7119  0.2654   
1    0.08690  0.07017  ...  158.80  1956.0  0.12380  0.18660  0.2416  0.1860   
2    0.19740  0.12790  ...  152.50  1709.0  0.14440  0.42450  0.4504  0.2430   
4    0.19800  0.10430  ...  152.20  1575.0  0.13740  0.20500  0.4000  0.1625   
6    0.11270  0.07400  ...  153.20  1606.0  0.14420  0.25760  0.3784  0.1932   
..       ...      ...  ...     ...     ...      ...      ...     ...     ...   
564  0.24390  0.13890  ...  166.10  2027.0  0.14100  0.21130  0.4107  0.2216   
565  0.14400  0.09791  ...  155.00  1731.0  0.11660  0.19220  0.3215  0.1628   
566  0.09251  0.05302  ...  126.70  1124.0  0.11390  0.30940  0.3403  0.1418   
567  0.35140  0.15200  ...  184.60  1821.0  0.16500  0.86810  0.9387  0.2650   
568  0.00000  0.00000  ...   59.16   268.6  0.08996  0.06444  0.0000  0.0000   

        f29      f30  target  label_predicted  
0    0.4601  0.11890       1                1  
1    0.2750  0.08902       1                1  
2    0.3613  0.08758       1                1  
4    0.2364  0.07678       1                1  
6    0.3063  0.08368       1                1  
..      ...      ...     ...              ...  
564  0.2060  0.07115       1                1  
565  0.2572  0.06637       1                1  
566  0.2218  0.07820       1                1  
567  0.4087  0.12400       1                1  
568  0.2871  0.07039       0                0  

[507 rows x 34 columns]

**Predict the Probability of Each Cell Type in Training Set**

In [183]:
model.predict_proba(x_tr)[:5,:]

array([[0.00893677, 0.99106323],
       [0.00135816, 0.99864184],
       [0.95568155, 0.04431845],
       [0.7259081 , 0.2740919 ],
       [0.93858145, 0.06141855]])

For example, for the first subject, the probability of a benign cell is 0.99106323
The probability of a malignant cell is 0.99018766

Since the malignant probability is greater than the default threshold of 0.5, the predicted cell type is 1 (malignant).

**Extracting Positive Probabilities**

It can be useful to extract the probabilities of the positive label for each subject.  
This can be done by extracting the second index across all rows like this:

In [184]:
model.predict_proba(x_tr)[:,1]

array([9.91063235e-01, 9.98641837e-01, 4.43184472e-02, 2.74091901e-01,
       6.14185475e-02, 6.67955381e-01, 5.42011809e-04, 9.98889274e-01,
       2.88187797e-02, 1.95646879e-04, 1.92967094e-01, 9.71140017e-01,
       9.97634959e-01, 7.26186116e-01, 1.80975222e-01, 9.63717313e-01,
       9.19852744e-01, 9.99284578e-01, 5.28076017e-01, 6.07927249e-04,
       3.83881847e-02, 2.65537551e-01, 4.02584320e-02, 9.93553613e-01,
       4.07013743e-02, 9.74593533e-01, 9.96214999e-01, 2.53165292e-01,
       9.99693326e-01, 9.55925072e-02, 9.99761096e-01, 7.01868445e-01,
       4.28926102e-02, 9.74549370e-01, 2.64989478e-02, 1.58036169e-01,
       1.24316415e-01, 3.73227274e-02, 5.60511843e-02, 3.73205023e-01,
       6.85587717e-01, 5.90320061e-01, 9.60803243e-03, 2.33751855e-02,
       9.86887680e-01, 9.54189030e-01, 9.99686449e-01, 6.68022413e-03,
       7.44899193e-02, 1.61047431e-02, 2.47894574e-02, 9.79192148e-01,
       9.77291608e-03, 7.41325736e-01, 8.69464391e-02, 8.15890955e-01,
      

Suppose we want to change the threshold, predicting cell type 1 if the probability of the positive label is greater than 0.85. This now requires a higher confidence compared to using 0.5 as the threshold.

In [186]:
model.predict_proba(x_tr)[:,1] > 0.85

array([ True,  True, False, False, False, False, False,  True, False,
       False, False,  True,  True, False, False,  True,  True,  True,
       False, False, False, False, False,  True, False,  True,  True,
       False,  True, False,  True, False, False,  True, False, False,
       False, False, False, False, False, False, False, False,  True,
        True,  True, False, False, False, False,  True, False, False,
       False, False,  True, False, False, False,  True,  True, False,
       False,  True, False, False,  True, False, False, False,  True,
        True,  True, False, False, False, False, False, False, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False, False, False, False, False, False,  True, False, False,
       False, False, False, False, False,  True,  True, False, False,
       False, False,  True, False, False,  True, False, False, False,
       False, False, False,  True, False,  True, False,  True, False,
       False, False,

**Measuring Performance**

In [188]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import precision_recall_curve

In [189]:
confusion_matrix(y_te, model.predict(x_te))

array([[141,  10],
       [ 19,  58]], dtype=int64)

Compute batch of metrics

In [190]:
print(classification_report(y_te, model.predict(x_te)))

              precision    recall  f1-score   support

           0       0.88      0.93      0.91       151
           1       0.85      0.75      0.80        77

    accuracy                           0.87       228
   macro avg       0.87      0.84      0.85       228
weighted avg       0.87      0.87      0.87       228



Compute metrics over range of thresholds

In [191]:
# probabilities of positive class from test set
pred_pos = model.predict_proba(x_te)[:,1]

# calculate metrics over range of thresholds
pre, re, th = precision_recall_curve(y_te,  pred_pos)

# print metrics for a sample of thresholds
print('precision:', pre[100:105])
print('')
print('recall:   ', re[100:105])
print('')
print('threshold:', th[100:105])

precision: [0.87692308 0.890625   0.9047619  0.90322581 0.90163934]

recall:    [0.74025974 0.74025974 0.74025974 0.72727273 0.71428571]

threshold: [0.52215923 0.53424723 0.54119652 0.55717661 0.56369838]


---

**THINK ABOUT AND DISCUSS**

1) If you raise the threshold for predicting a positive label, what generally happens to the recall? What generally happens to the precision?

answer: Since we require a higher confidence to predict the positive label, the precision will likely increase.  However, the recall will likely decrease, producing more false negatives. This is why it is important to measure precision and recall together.

---

**TRY FOR YOURSELF**

You will evaluate the model performance.

Hint: you can count the number of rows in a dataframe by using the `len()` function like this: `len(df)`

2) Compute the accuracy of the model, where accuracy = #correct / #total

In [ ]:
# answer
len(df[df['label_predicted'] == df['target']]) / len(df)

3) Count the number of true positives (where the model predicted 1 and the target is 1)

In [ ]:
# answer
len(df[ (df['label_predicted'] == 1) & (df['target'] == 1) ])

4) Compute the recall of the model, where precision = #true_positive / #actual_positive

In [ ]:
# answer
len(df[ (df['label_predicted'] == 1) & (df['target'] == 1) ]) / len( df[df['target'] == 1])

---

5) As we saw earlier, the first subject in the training set has a probability of malignancy = 0.99018766  
Compute this by using the intercept, coefficients, f1, f2 values, and the definition of the sigmoid:

$sigmoid = 1 / ( 1 + np.exp(-(b0 + b1 * x1 + b2 * x2) ))$

Note this version uses two predictors.

In [192]:
#answer
b0 = model.intercept_
b1 = model.coef_[0][0]
b2 = model.coef_[0][1]
x1 = x_tr[0][0]
x2 = x_tr[0][1]

1 / ( 1 + np.exp(-(b0 + b1 * x1 + b2 * x2) ))

array([0.99106323])

---